lib/tempfile.rb

```diff
@@ -464,3 +464,89 @@ def Tempfile.create(basename="", tmpdir=nil, mode: 0, **options)
     tmpfile
   end
 end
+
+# Creates an unnamed temporary file in the underlying file system;
+# returns a new \IO object based on that file.
+#
+# This method removes the file before returning the \IO object.
+# Thus, no need to remove the file.
+#
+# Argument +basename+, +tmpdir+, keyword arguments +mode+ and +options+ are same as
+# Tempfile.create.
+#
+# This method may take a block.
+# If the block is not given, the created \IO object is returned.
+# The caller should close it.
+# If the block is given, the created \IO object is yielded.
+# It is closed afeter the block is finished.
+#
+# With no arguments, creates an \IO object whose:
+#
+# - Class is {IO}[rdoc-ref:IO] (not \Tempfile).
+# - Directory is the system temporary directory (system-dependent).
+# - Permissions are <tt>0600</tt>;
+#   see {File Permissions}[rdoc-ref:File@File+Permissions].
+# - Mode is <tt>'w+'</tt> (read/write mode, positioned at the end).
+#
+# Example:
+#
+#   Tempfile.create_io {|tmpio|
+#     tmpio.puts "foo"
+#     tmpio.rewind
+#     p tmpio.read              # => "foo\n"
+#   }
+#
+#   tmpio = Tempfile.create_io  # => #<IO:fd 5>
+#   tmpio.class                 # => IO
+#   tmpio.path                  # => nil
+#   tmpio.stat.mode.to_s(8)     # => "100600"
+#   tmpio.puts "foo"
+#   tmpio.rewind
+#   tmpio.read                  # => "foo\n"
+#   tmpio.close
+#
+#
+# Implementation:
+#
+# On POSIX systems, Tempfile.create_io calls Tempfile.create and the created
+# file is unlinked.
+# Thus Tempfile.create_io needs arguments same as Tempfile.create.
+#
+# On Linux, O_TMPFILE flag is used to create an unnamed file.
+# O_TMPFILE makes possible to create a file without filename.
+# The argument +tmpdir+ is still used to create the unnamed file.
+# If the filesystem which is specified by +tmpdir+ doesn't support O_TMPFILE,
+# this method fallbacks to create-and-unlink as on POSIX.
+#
+# Related: Tempfile.create.
+def Tempfile.create_io(basename="", tmpdir=nil, mode: 0, **options, &block)
+  tmpio = nil
+  if defined? File::TMPFILE # O_TMPFILE since Linux 3.11
+    tmpfile_supported = true
+    tmpdir = Dir.tmpdir() if tmpdir.nil?
+    begin
+      fd = IO.sysopen(tmpdir, File::RDWR | File::TMPFILE, 0600)
+    rescue Errno::EISDIR, Errno::ENOENT, Errno::EOPNOTSUPP # kernel or the filesystem does not support O_TMPFILE
+      tmpfile_supported = false
+    end
+    if tmpfile_supported
+      tmpio = IO.new(fd, File::RDWR)
+    end
+  end
+  if tmpio.nil?
+    mode |= File::SHARE_DELETE
```

Comment:
In `w32_wopen`, `_wopen` is used unless binary mode, and it does not have share mode argument.
```suggestion
    mode |= File::SHARE_DELETE | File::BINARY
```